export/securedrop_export/print/service.py

```diff
@@ -210,53 +232,153 @@ def _setup_printer(self, printer_uri, printer_ppd):
+        """
+        supported_mimetypes: set[str] = set()
+        libreoffice_programs = [
+            "libreoffice-base",
+            "libreoffice-calc",
+            "libreoffice-draw",
+            "libreoffice-impress",
+            "libreoffice-math",
+            "libreoffice-writer",
         ]
-        for extension in OPEN_OFFICE_FORMATS:
-            if os.path.basename(filename).endswith(extension):
-                return True
-        return False
+        for item in libreoffice_programs:
+            desktop_file = Path(desktop_dir, f"{item}.desktop")
+            if desktop_file.exists():
+                with open(desktop_file) as f:
+                    for line in f.readlines():
+                        if line.startswith("MimeType="):
+                            # Semicolon-separated list; don't leave empty element at the end
+                            supported_mimetypes.update(line.strip("MimeType=").split(";")[:-1])
+
+        return supported_mimetypes
+
+    def _needs_pdf_conversion(self, filename: Path):
+        """
+        Checks mimetype of a file and returns True if file must be converted
+        to PDF before attempting to print.
 
-    def _print_file(self, file_to_print):
-        # If the file to print is an (open)office document, we need to call unoconf to
-        # convert the file to pdf as printer drivers do not support this format
+        Raises ExportException for unprintable mimetypes or on mimetype
+        discovery error.
+        """
+        mimetype = None
+        supported_types: set[str] = set()
 
-        if self._is_open_office_file(file_to_print):
-            logger.info("Converting Office document to pdf")
-            folder = os.path.dirname(file_to_print)
-            converted_filename = file_to_print + ".pdf"
-            converted_path = os.path.join(folder, converted_filename)
+        try:
+            supported_types = self._get_supported_mimetypes_libreoffice(LIBREOFFICE_DESKTOP_DIR)
+        except OSError as e:
+            logger.error(f"Could not get supported mimetypes list: {e}")
+            raise ExportException(sdstatus=Status.ERROR_MIMETYPE_DISCOVERY)
+        if len(supported_types) == 0:
+            raise ExportException(sdstatus=Status.ERROR_MIMETYPE_DISCOVERY)
+
+        try:
+            # b'filename that may have spaces.docx: application/bla\n'
+            # use magic bytes (-M) for filetype detection
+            mimetype = (
+                subprocess.check_output(["mimetype", "-M", filename])
+                .decode()
+                .split(":")[-1]
+                .strip()
+            )
+        except subprocess.CalledProcessError:
+            logger.error(f"Could not process mimetype of {filename}")
+            raise ExportException(sdstatus=Status.ERROR_MIMETYPE_UNKNOWN)
+        # Don't print "audio/*", "video/*", or archive mimetypes
+        if (
+            any(mimetype.startswith(prefix) for prefix in MIMETYPE_UNPRINTABLE)
```

Comment:
A cool trick is that `startswith` takes a tuple of strings, which allows you to do:
```
>>> 'audio/foo'.startswith(('audio/', 'video/'))
True
>>> 'text/foo'.startswith(('audio/', 'video/'))
False
```